core/src/main/java/org/pentaho/di/core/database/Database.java

```diff
@@ -0,0 +1,5078 @@
+        // cannot do anything about this but log it
+        log.logError(
+          "Error disconnecting from database - closeConnectionOnly failed:" + Const.CR + ignoredKde.getMessage() );
+        log.logError( Const.getStackTracker( ignoredKde ) );
+      }
+    }
+  }
+
+  @Override public void close() {
+    disconnect();
+  }
+
+  /**
+   * Only for unique connections usage, typically you use disconnect() to disconnect() from the database.
+   *
+   * @throws KettleDatabaseException in case there is an error during connection close.
+   */
+  public synchronized void closeConnectionOnly() throws KettleDatabaseException {
+    try {
+      if ( connection != null ) {
+        connection.close();
+        if ( !databaseMeta.isUsingConnectionPool() ) {
+          connection = null;
+        }
+      }
+
+      if ( log.isDetailed() ) {
+        log.logDetailed( "Connection to database closed!" );
+      }
+    } catch ( SQLException e ) {
+      throw new KettleDatabaseException( "Error disconnecting from database '" + toString() + "'", e );
+    }
+  }
+
+  /**
+   * Cancel the open/running queries on the database connection
+   *
+   * @throws KettleDatabaseException
+   */
+  public void cancelQuery() throws KettleDatabaseException {
+    // Canceling statements only if we're not streaming results on MySQL with
+    // the v3 driver
+    //
+    if ( databaseMeta.isMySQLVariant()
+      && databaseMeta.isStreamingResults() && getDatabaseMetaData().getDriverMajorVersion() == 3 ) {
+      return;
+    }
+
+    cancelStatement( pstmt );
+    cancelStatement( selStmt );
+  }
+
+  /**
+   * Cancel an open/running SQL statement
+   *
+   * @param statement the statement to cancel
+   * @throws KettleDatabaseException
+   */
+  public void cancelStatement( Statement statement ) throws KettleDatabaseException {
+    try {
+      if ( statement != null ) {
+        statement.cancel();
+      }
+      if ( log.isDebug() ) {
+        log.logDebug( "Statement canceled!" );
+      }
+    } catch ( SQLException ex ) {
+      throw new KettleDatabaseException( "Error cancelling statement", ex );
+    }
+  }
+
+  /**
+   * Specify after how many rows a commit needs to occur when inserting or updating values.
+   *
+   * @param commsize The number of rows to wait before doing a commit on the connection.
+   */
+  public void setCommit( int commsize ) {
+    commitsize = commsize;
+    String onOff = ( commitsize <= 0 ? "on" : "off" );
+    try {
+      connection.setAutoCommit( commitsize <= 0 );
+      if ( log.isDetailed() ) {
+        log.logDetailed( "Auto commit " + onOff );
+      }
+    } catch ( Exception e ) {
+      if ( log.isDebug() ) {
+        log.logDebug( "Can't turn auto commit " + onOff + Const.CR + Const.getStackTracker( e ) );
+      }
+    }
+  }
+
+  public void setAutoCommit( boolean useAutoCommit ) throws KettleDatabaseException {
+    try {
+      connection.setAutoCommit( useAutoCommit );
+    } catch ( SQLException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)